File: projects/basis/basis/modules/private_markets.py
import logging
from web3 import Web3
from .factory import load_abi

logger = logging.getLogger(__name__)


class PrivateMarketsModule:
    def __init__(self, client, private_market_address: str):
        self.client = client
        self.private_market_address = Web3.to_checksum_address(private_market_address)
        self.private_market_abi = load_abi('APrivateTradingMarket.json')
        self.erc20_abi = load_abi('IERC20.json')
        self.contract = self.client.web3.eth.contract(address=self.private_market_address, abi=self.private_market_abi)

    def _approve_if_needed(self, token_address: str, amount: int):
        if not self.client.account:
            raise ValueError("Wallet account is required for approval.")

        checksum_addr = Web3.to_checksum_address(token_address)
        token_contract = self.client.web3.eth.contract(address=checksum_addr, abi=self.erc20_abi)

        allowance = token_contract.functions.allowance(
            self.client.account.address, self.private_market_address
        ).call()

        if allowance < amount:
            func = token_contract.functions.approve(self.private_market_address, amount)
            tx = func.build_transaction({
                'from': self.client.account.address,
                'nonce': self.client.web3.eth.get_transaction_count(self.client.account.address),
            })
            signed_tx = self.client.web3.eth.account.sign_transaction(tx, private_key=self.client.account.key)
            tx_hash = self.client.web3.eth.send_raw_transaction(signed_tx.raw_transaction)
            self.client.web3.eth.wait_for_transaction_receipt(tx_hash)

    def _build_and_send_tx(self, function_call, value=0):
        if not self.client.account:
            raise ValueError("Stateful initialization (private_key) is required for write methods.")

        tx = function_call.build_transaction({
            'from': self.client.account.address,
            'nonce': self.client.web3.eth.get_transaction_count(self.client.account.address),
            'value': value,
        })

        signed_tx = self.client.web3.eth.account.sign_transaction(tx, private_key=self.client.account.key)
        tx_hash = self.client.web3.eth.send_raw_transaction(signed_tx.raw_transaction)
        receipt = self.client.web3.eth.wait_for_transaction_receipt(tx_hash)

        return {
            'hash': tx_hash.hex(),
            'receipt': receipt
        }

    def _sync_order(self, tx_hash: str):
        """Sync order tx to backend. Non-fatal on failure."""
        try:
            if not tx_hash.startswith("0x"):
                tx_hash = "0x" + tx_hash
            self.client.api.sync_order(tx_hash, 'private')
        except Exception as e:
            logger.warning("Order sync warning: %s", e)

    # ------------------------------------------------------------------
    # Write methods
    # ------------------------------------------------------------------

    def create_market(self, market_name: str, symbol: str, end_time: int, option_names: list[str], maintoken: str, frozen: bool, bonding: int, seed_amount: int = 0):
        """Creates a private prediction market. Auto-fetches and attaches the creation fee."""
        checksum_maintoken = Web3.to_checksum_address(maintoken)
        eco_data = self.contract.functions.ecosystems(checksum_maintoken).call()
        factory_address = eco_data[0]
        factory_abi = load_abi('ATokenFactory.json')
        factory_contract = self.client.web3.eth.contract(address=factory_address, abi=factory_abi)
        fee_amount = factory_contract.functions.feeAmount().call()

        # Auto-approve USDB for seed amount if needed
        if seed_amount > 0:
            self._approve_if_needed(self.client.usdb_address, seed_amount)

        func = self.contract.functions.createMarket(market_name, symbol, end_time, option_names, checksum_maintoken, frozen, bonding, seed_amount)
        if not self.client.account:
            raise ValueError("Stateful initialization (private_key) is required for write methods.")
        tx = func.build_transaction({
            'from': self.client.account.address,
            'nonce': self.client.web3.eth.get_transaction_count(self.client.account.address),
            'value': fee_amount,
        })
        signed_tx = self.client.web3.eth.account.sign_transaction(tx, private_key=self.client.account.key)
        tx_hash = self.client.web3.eth.send_raw_transaction(signed_tx.raw_transaction)
        receipt = self.client.web3.eth.wait_for_transaction_receipt(tx_hash)
        return {'hash': tx_hash.hex(), 'receipt': receipt}

    def buy(self, market_token: str, outcome_id: int, input_token: str, input_amount: int, min_usdb: int, min_shares: int):
        """Buys shares in a private market outcome. Auto-approves input token."""
        checksum_market = Web3.to_checksum_address(market_token)
        checksum_input = Web3.to_checksum_address(input_token)
        self._approve_if_needed(checksum_input, input_amount)
        func = self.contract.functions.buy(checksum_market, outcome_id, checksum_input, input_amount, min_usdb, min_shares)
        return self._build_and_send_tx(func)

    def redeem(self, market_token: str):
        """Redeems shares from a resolved private market."""
        checksum_market = Web3.to_checksum_address(market_token)
        func = self.contract.functions.redeem(checksum_market)
        return self._build_and_send_tx(func)

    def list_order(self, market_token: str, outcome_id: int, amount: int, price_per_share: int):
        """Lists a sell order on the private market order book."""
        checksum_market = Web3.to_checksum_address(market_token)
        func = self.contract.functions.listOrder(checksum_market, outcome_id, amount, price_per_share)
        result = self._build_and_send_tx(func)
        self._sync_order(result['hash'])
        return result

    def cancel_order(self, market_token: str, order_id: int):
        """Cancels an existing order on the private market."""
        checksum_market = Web3.to_checksum_address(market_token)
        func = self.contract.functions.cancelOrder(checksum_market, order_id)
        result = self._build_and_send_tx(func)
        self._sync_order(result['hash'])
        return result

    def buy_order(self, market_token: str, order_id: int, fill: int):
        """Buys from an existing order on the private market."""
        checksum_market = Web3.to_checksum_address(market_token)
        func = self.contract.functions.buyOrder(checksum_market, order_id, fill)
        result = self._build_and_send_tx(func)
        self._sync_order(result['hash'])
        return result

    def buy_multiple_orders(self, market_token: str, order_ids: list[int], usdb_amount: int):
        """Buys from multiple orders on the private market."""
        checksum_market = Web3.to_checksum_address(market_token)
        func = self.contract.functions.buyMultipleOrders(checksum_market, order_ids, usdb_amount)
        result = self._build_and_send_tx(func)
        self._sync_order(result['hash'])
        return result

    def buy_orders_and_contract(self, market_token: str, outcome_id: int, order_ids: list[int], input_token: str, total_input: int, min_shares: int):
        """Buys from order book and AMM in a single transaction. Auto-approves input token."""
        checksum_market = Web3.to_checksum_address(market_token)
        checksum_input = Web3.to_checksum_address(input_token)
        self._approve_if_needed(checksum_input, total_input)
        func = self.contract.functions.buyOrdersAndContract(checksum_market, outcome_id, order_ids, checksum_input, total_input, min_shares)
        result = self._build_and_send_tx(func)
        self._sync_order(result['hash'])
        return result

    # ------------------------------------------------------------------
    # Read methods
    # ------------------------------------------------------------------

    def get_market_data(self, market_token: str):
        """Returns market data for a private market."""
        checksum_market = Web3.to_checksum_address(market_token)
        return self.contract.functions.getMarketData(checksum_market).call()

    def get_outcome(self, market_token: str, outcome_id: int):
        """Returns outcome data for a specific outcome."""
        checksum_market = Web3.to_checksum_address(market_token)
        return self.contract.functions.getOutcome(checksum_market, outcome_id).call()

    def get_user_shares(self, market_token: str, user: str, outcome_id: int):
        """Returns a user's shares for a specific outcome."""
        checksum_market = Web3.to_checksum_address(market_token)
        checksum_user = Web3.to_checksum_address(user)
        return self.contract.functions.getUserShares(checksum_market, checksum_user, outcome_id).call()

    def get_buy_order_cost(self, market_token: str, order_id: int, fill: int):
        """Returns the cost to buy a specific order fill amount."""
        checksum_market = Web3.to_checksum_address(market_token)
        return self.contract.functions.getBuyOrderCost(checksum_market, order_id, fill).call()

    def get_initial_reserves(self, num_outcomes: int) -> int:
        """Returns the initial reserves required for a given number of outcomes."""
        return self.contract.functions.getInitialReserves(num_outcomes).call()

File: projects/basis/sdk-review/py-v2/src/modules/PrivateMarkets.ts
import { BasisClient } from '../BasisClient';
import APrivateTradingMarketArtifact from '../abis/APrivateTradingMarket.json';
import ATokenFactoryArtifact from '../abis/ATokenFactory.json';
import IERC20Artifact from '../abis/IERC20.json';
import { Address } from 'viem';

export class PrivateMarketsModule {
  private client: BasisClient;
  private privateMarketAddress: Address;

  constructor(client: BasisClient, privateMarketAddress: Address) {
    this.client = client;
    this.privateMarketAddress = privateMarketAddress;
  }

  private async approveIfNeeded(tokenAddress: Address, spender: Address, amount: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Wallet account is required for approval.");
    }
    const account = this.client.walletClient.account;
    const allowance = await this.client.publicClient.readContract({
      address: tokenAddress,
      abi: IERC20Artifact.abi,
      functionName: 'allowance',
      args: [account.address, spender],
    }) as bigint;

    if (allowance < amount) {
      const { request } = await this.client.publicClient.simulateContract({
        account,
        address: tokenAddress,
        abi: IERC20Artifact.abi,
        functionName: 'approve',
        args: [spender, amount],
      });
      const hash = await this.client.walletClient.writeContract(request);
      await this.client.publicClient.waitForTransactionReceipt({ hash });
    }
  }

  private async _syncTx(txHash: string) {
    try {
      await this.client.api.syncTransaction(txHash);
    } catch (e: any) {
      console.warn('Sync warning:', e.message || e);
    }
  }

  private async syncOrder(txHash: string): Promise<void> {
    try {
      await this.client.api.syncOrder(txHash, 'private');
    } catch (err) {
      console.warn('Order sync warning:', err instanceof Error ? err.message : err);
    }
  }

  // -----------------------------------------------------------------------
  // Write methods
  // -----------------------------------------------------------------------

  /**
   * Creates a new private prediction market.
   * Fetches the ecosystem factory fee and attaches it.
   */
  async createMarket(
    marketName: string,
    symbol: string,
    endTime: bigint,
    optionNames: string[],
    maintoken: Address,
    privateEvent: boolean,
    frozen: boolean,
    bonding: bigint,
    seedAmount: bigint = 0n
  ) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    // Get the ecosystem's factory address, then fetch the fee
    const ecoData = await this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'ecosystems',
      args: [maintoken],
    }) as any;
    const factoryAddress = ecoData.factory ?? ecoData[0];
    const feeAmount = await this.client.publicClient.readContract({
      address: factoryAddress,
      abi: ATokenFactoryArtifact.abi,
      functionName: 'feeAmount',
    }) as bigint;

    // Auto-approve USDB for seed amount if needed
    if (seedAmount > 0n) {
      await this.approveIfNeeded(this.client.usdbAddress, this.privateMarketAddress, seedAmount);
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'createMarket',
      args: [marketName, symbol, endTime, optionNames, maintoken, privateEvent, frozen, bonding, seedAmount],
      value: feeAmount,
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    await this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Executes an AMM buy for a private market outcome.
   * Auto-approves the input token.
   */
  async buy(
    marketToken: Address,
    outcomeId: number,
    inputToken: Address,
    inputAmount: bigint,
    minUsdb: bigint,
    minShares: bigint
  ) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    await this.approveIfNeeded(inputToken, this.privateMarketAddress, inputAmount);

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'buy',
      args: [marketToken, outcomeId, inputToken, inputAmount, minUsdb, minShares],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    await this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Redeems winnings after a market resolves.
   */
  async redeem(marketToken: Address) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'redeem',
      args: [marketToken],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    await this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Creates a limit order on a private market.
   */
  async listOrder(marketToken: Address, outcomeId: number, amount: bigint, pricePerShare: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'listOrder',
      args: [marketToken, outcomeId, amount, pricePerShare],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    await this.syncOrder(hash);

    return { hash, receipt };
  }

  /**
   * Cancels an active order on a private market.
   */
  async cancelOrder(marketToken: Address, orderId: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'cancelOrder',
      args: [marketToken, orderId],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    await this.syncOrder(hash);

    return { hash, receipt };
  }

  /**
   * Fills a specific order on a private market.
   */
  async buyOrder(marketToken: Address, orderId: bigint, fill: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'buyOrder',
      args: [marketToken, orderId, fill],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    await this.syncOrder(hash);

    return { hash, receipt };
  }

  /**
   * Sweeps multiple orders on a private market.
   */
  async buyMultipleOrders(marketToken: Address, orderIds: bigint[], usdbAmount: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'buyMultipleOrders',
      args: [marketToken, orderIds, usdbAmount],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    await this.syncOrder(hash);

    return { hash, receipt };
  }

  /**
   * Buys from order book and AMM in a single transaction.
   * Auto-approves the input token.
   */
  async buyOrdersAndContract(
    marketToken: Address,
    outcomeId: number,
    orderIds: bigint[],
    inputToken: Address,
    totalInput: bigint,
    minShares: bigint
  ) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    await this.approveIfNeeded(inputToken, this.privateMarketAddress, totalInput);

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'buyOrdersAndContract',
      args: [marketToken, outcomeId, orderIds, inputToken, totalInput, minShares],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    await this.syncOrder(hash);

    return { hash, receipt };
  }

  /**
   * Casts a vote on a private market outcome.
   */
  async vote(marketToken: Address, outcomeId: number) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'vote',
      args: [marketToken, outcomeId],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    await this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Finalizes a private market after voting is complete.
   */
  async finalize(marketToken: Address) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'finalize',
      args: [marketToken],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    await this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Claims the bounty reward for voting correctly.
   */
  async claimBounty(marketToken: Address) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'claimBounty',
      args: [marketToken],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    await this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Manages voter status for a private market.
   */
  async manageVoter(marketToken: Address, voter: Address, status: boolean) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'manageVoter',
      args: [marketToken, voter, status],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    await this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Toggles whether specific addresses can buy in a private event market.
   */
  async togglePrivateEventBuyers(marketToken: Address, buyers: Address[], status: boolean) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'togglePrivateEventBuyers',
      args: [marketToken, buyers, status],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    await this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Disables the freeze on a private market.
   */
  async disableFreeze(marketToken: Address) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'DisableFreeze',
      args: [marketToken],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    await this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Manages the whitelist for a private market.
   */
  async manageWhitelist(marketToken: Address, wallets: Address[], amount: bigint, tag: string, status: boolean) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'manageWhitelist',
      args: [marketToken, wallets, amount, tag, status],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    await this._syncTx(hash);

    return { hash, receipt };
  }

  // -----------------------------------------------------------------------
  // Read methods
  // -----------------------------------------------------------------------

  /**
   * Returns the MarketData struct for a private market.
   */
  async getMarketData(marketToken: Address) {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'getMarketData',
      args: [marketToken],
    });
  }

  /**
   * Returns the number of outcomes for a market.
   */
  async getNumOutcomes(marketToken: Address): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'getNumOutcomes',
      args: [marketToken],
    }) as Promise<bigint>;
  }

  /**
   * Returns the Outcome struct for a specific outcome.
   */
  async getOutcome(marketToken: Address, outcomeId: bigint) {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'outcomes',
      args: [marketToken, outcomeId],
    });
  }

  /**
   * Returns user shares for a specific outcome.
   */
  async getUserShares(marketToken: Address, user: Address, outcomeId: number): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'userShares',
      args: [marketToken, user, outcomeId],
    }) as Promise<bigint>;
  }

  /**
   * Returns whether a user has bet on a market.
   */
  async hasBetted(marketToken: Address, user: Address): Promise<boolean> {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'hasBetted',
      args: [marketToken, user],
    }) as Promise<boolean>;
  }

  /**
   * Returns the bounty pool amount for a market.
   */
  async getBountyPool(marketToken: Address): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'bountyPool',
      args: [marketToken],
    }) as Promise<bigint>;
  }

  /**
   * Returns the cost to buy an order.
   */
  async getBuyOrderCost(marketToken: Address, orderId: bigint, fill: bigint) {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'getBuyOrderCost',
      args: [marketToken, orderId, fill],
    });
  }

  /**
   * Returns the amounts out when buying an order with a specific USDB amount.
   */
  async getBuyOrderAmountsOut(marketToken: Address, orderId: bigint, usdbAmount: bigint) {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'getBuyOrderAmountsOut',
      args: [marketToken, orderId, usdbAmount],
    });
  }

  /**
   * Returns an order by market and order ID.
   */
  async getMarketOrders(marketToken: Address, orderId: bigint) {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'marketOrders',
      args: [marketToken, orderId],
    });
  }

  /**
   * Returns the next order ID for a market.
   */
  async getNextOrderId(marketToken: Address): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'nextOrderId',
      args: [marketToken],
    }) as Promise<bigint>;
  }

  /**
   * Returns whether an address is a voter for a market.
   */
  async isMarketVoter(marketToken: Address, voter: Address): Promise<boolean> {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'isMarketVoter',
      args: [marketToken, voter],
    }) as Promise<boolean>;
  }

  /**
   * Returns the outcome a voter chose for a market.
   */
  async getVoterChoice(marketToken: Address, voter: Address): Promise<number> {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'voterChoice',
      args: [marketToken, voter],
    }) as Promise<number>;
  }

  /**
   * Returns the first vote time for a market.
   */
  async getFirstVoteTime(marketToken: Address): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'firstVoteTime',
      args: [marketToken],
    }) as Promise<bigint>;
  }

  /**
   * Returns whether a user can buy in a private event market.
   */
  async canUserBuy(marketToken: Address, user: Address): Promise<boolean> {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'userCanBuyEvent',
      args: [marketToken, user],
    }) as Promise<boolean>;
  }

  /**
   * Returns the bounty per correct vote for a market.
   */
  async getBountyPerVote(marketToken: Address): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'bountyPerCorrectVote',
      args: [marketToken],
    }) as Promise<bigint>;
  }

  /**
   * Returns whether a voter has claimed the bounty for a market.
   */
  async hasClaimed(marketToken: Address, voter: Address): Promise<boolean> {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'bountyClaimed',
      args: [marketToken, voter],
    }) as Promise<boolean>;
  }

  /**
   * Returns the initial reserves required for a given number of outcomes.
   */
  async getInitialReserves(numOutcomes: bigint): Promise<readonly [bigint, bigint]> {
    return this.client.publicClient.readContract({
      address: this.privateMarketAddress,
      abi: APrivateTradingMarketArtifact.abi,
      functionName: 'getInitialReserves',
      args: [numOutcomes],
    }) as Promise<readonly [bigint, bigint]>;
  }
}
